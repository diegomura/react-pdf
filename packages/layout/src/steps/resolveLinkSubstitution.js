import * as P from '@react-pdf/primitives';
import { compose } from '@react-pdf/fns';

const isType = type => node => node.type === type;

const isLink = isType(P.Link);

const isText = isType(P.Text);

const isTextInstance = isType(P.TextInstance);

/**
 * Checks if node has render prop
 *
 * @param {Object} node
 * @returns {Boolean} has render prop?
 */
const hasRenderProp = node => !!node.props?.render;

/**
 * Checks if node is text type (Text or TextInstance)
 *
 * @param {Object} node
 * @returns {Boolean} are all children text instances?
 */
const isTextType = node => isText(node) || isTextInstance(node);

/**
 * Checks if is tet link that needs to be wrapped in Text
 *
 * @param {Object} node
 * @returns {Boolean} are all children text instances?
 */
const isTextLink = node => {
  const children = node.children || [];

  // Text string inside a Link
  if (children.every(isTextInstance)) return true;

  // Text node inside a Link
  if (children.every(isText)) return false;

  return children.every(isTextType);
};

/**
 * Wraps node children inside Text node
 *
 * @param {Object} node
 * @returns {Boolean} node with intermediate Text child
 */
const wrapText = node => {
  const textElement = {
    type: P.Text,
    props: {},
    style: {},
    box: {},
    children: node.children,
  };

  return Object.assign({}, node, { children: [textElement] });
};

const transformLink = node => {
  if (!isLink(node)) return node;

  // If has render prop substitute the instance by a Text, that will
  // ultimately render the inline Link via the textkit PDF renderer.
  if (hasRenderProp(node)) return Object.assign({}, node, { type: P.Text });

  // If is a text link (either contains Text or TextInstalce), wrap it
  // inside a Text element so styles are applied correctly

  if (isTextLink(node)) return wrapText(node);

  return node;
};

/**
 * Transforms Link layout to correctly render text and dynamic rendered links
 *
 * @param {Object} node
 * @returns {Object} node with link substitution
 */
const resolveLinkSubstitution = node => {
  if (!node.children) return node;

  const resolveChild = compose(transformLink, resolveLinkSubstitution);

  const children = node.children.map(resolveChild);

  return Object.assign({}, node, { children });
};

export default resolveLinkSubstitution;
