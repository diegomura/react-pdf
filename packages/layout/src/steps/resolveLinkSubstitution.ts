import * as P from '@react-pdf/primitives';
import { compose } from '@react-pdf/fns';

import { Node } from '../types';

const isType = (type: string) => (node: Node) => node.type === type;

const isLink = isType(P.Link);

const isText = isType(P.Text);

const isTextInstance = isType(P.TextInstance);

/**
 * Checks if node has render prop
 *
 * @param node
 * @returns Has render prop?
 */
const hasRenderProp = (node: Node) => 'render' in node.props;

/**
 * Checks if node is text type (Text or TextInstance)
 *
 * @param node
 * @returns Are all children text instances?
 */
const isTextType = (node: Node) => isText(node) || isTextInstance(node);

/**
 * Checks if is tet link that needs to be wrapped in Text
 *
 * @param node
 * @returns Are all children text instances?
 */
const isTextLink = (node: Node) => {
  const children = node.children || [];

  // Text string inside a Link
  if (children.every(isTextInstance)) return true;

  // Text node inside a Link
  if (children.every(isText)) return false;

  return children.every(isTextType);
};

/**
 * Wraps node children inside Text node
 *
 * @param node
 * @returns Node with intermediate Text child
 */
const wrapText = (node) => {
  const textElement = {
    type: P.Text,
    props: {},
    style: {},
    box: {},
    children: node.children,
  };

  return Object.assign({}, node, { children: [textElement] });
};

const transformLink = (node: Node) => {
  if (!isLink(node)) return node;

  // If has render prop substitute the instance by a Text, that will
  // ultimately render the inline Link via the textkit PDF renderer.
  if (hasRenderProp(node)) return Object.assign({}, node, { type: P.Text });

  // If is a text link (either contains Text or TextInstance), wrap it
  // inside a Text element so styles are applied correctly

  if (isTextLink(node)) return wrapText(node);

  return node;
};

/**
 * Transforms Link layout to correctly render text and dynamic rendered links
 *
 * @param node
 * @returns Node with link substitution
 */
const resolveLinkSubstitution = (node: Node): Node => {
  if (!node.children) return node;

  const resolveChild = compose(transformLink, resolveLinkSubstitution);

  const children = node.children.map(resolveChild);

  return Object.assign({}, node, { children });
};

export default resolveLinkSubstitution;
