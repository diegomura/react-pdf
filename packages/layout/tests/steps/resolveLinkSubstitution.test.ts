import { describe, expect, test } from 'vitest';

import resolveLinkSubstitution from '../../src/steps/resolveLinkSubstitution';

describe('layout resolve link substitution', () => {
  test('Should leave link with text children as it is', () => {
    const result = resolveLinkSubstitution({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          children: [
            {
              type: 'LINK',
              props: { src: 'url' },
              children: [
                {
                  type: 'TEXT',
                  props: {},
                },
              ],
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should replace link with only one text instance as children', () => {
    const result = resolveLinkSubstitution({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          children: [
            {
              type: 'LINK',
              props: { src: 'url' },
              children: [{ type: 'TEXT_INSTANCE', value: '1' }],
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should replace link with only many text instances as children', () => {
    const result = resolveLinkSubstitution({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          children: [
            {
              type: 'LINK',
              props: { src: 'url' },
              children: [
                { type: 'TEXT_INSTANCE', value: '1' },
                { type: 'TEXT_INSTANCE', value: '2' },
                { type: 'TEXT_INSTANCE', value: '3' },
                { type: 'TEXT_INSTANCE', value: '4' },
              ],
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should replace link with render prop', () => {
    const result = resolveLinkSubstitution({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          children: [
            {
              type: 'LINK',
              props: {
                render: () => null,
              },
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });
});
