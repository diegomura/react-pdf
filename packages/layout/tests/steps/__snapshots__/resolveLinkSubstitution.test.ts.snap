// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout resolve link substitution > Should leave link with text children as it is 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "props": {},
              "type": "TEXT",
            },
          ],
          "props": {
            "src": "url",
          },
          "type": "LINK",
        },
      ],
      "props": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolve link substitution > Should replace link with only many text instances as children 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "box": {},
              "children": [
                {
                  "type": "TEXT_INSTANCE",
                  "value": "1",
                },
                {
                  "type": "TEXT_INSTANCE",
                  "value": "2",
                },
                {
                  "type": "TEXT_INSTANCE",
                  "value": "3",
                },
                {
                  "type": "TEXT_INSTANCE",
                  "value": "4",
                },
              ],
              "props": {},
              "style": {},
              "type": "TEXT",
            },
          ],
          "props": {
            "src": "url",
          },
          "type": "LINK",
        },
      ],
      "props": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolve link substitution > Should replace link with only one text instance as children 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "box": {},
              "children": [
                {
                  "type": "TEXT_INSTANCE",
                  "value": "1",
                },
              ],
              "props": {},
              "style": {},
              "type": "TEXT",
            },
          ],
          "props": {
            "src": "url",
          },
          "type": "LINK",
        },
      ],
      "props": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolve link substitution > Should replace link with render prop 1`] = `
{
  "children": [
    {
      "children": [
        {
          "props": {
            "render": [Function],
          },
          "type": "TEXT",
        },
      ],
      "props": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;
